packages/dev/core/src/Behaviors/Meshes/fadeInOutBehavior.ts

```diff
@@ -58,28 +65,56 @@ export class FadeInOutBehavior implements Behavior<Mesh> {
      * @param value if the object should fade in or out (true to fade in)
      */
     public fadeIn(value: boolean) {
+        // Otherwise the value has changed so abort the next update
+        if (this._timeoutHandle) {
+            // prevent any pending updates
+            clearTimeout(this._timeoutHandle);
+            this._timeoutHandle = null;
+        }
+
+        if (this._ownerNode && 
+            ((value && this._ownerNode.visibility >= 1) || 
+            (!value && this._ownerNode.visibility <= 0))) {
+            // If fading in and already visible or fading out and already not visible do nothing
+            return;
+        }
+
         this._hovered = value;
+
+        // Reset the hoverValue.  This is neccessary becasue we may have been fading out, e.g. but not yet reached
+        // the delay, so the hover value is greater than 1
+        if (this._ownerNode!.visibility >= 1) {
+            this._hoverValue = this.fadeInTime;
+        } else if (this._ownerNode!.visibility <= 0) {
+            this._hoverValue = 0;
+        }
         this._update();
     }
 
     private _update = () => {
         if (this._ownerNode) {
             this._hoverValue += this._hovered ? this._millisecondsPerFrame : -this._millisecondsPerFrame;
 
-            this._setAllVisibility(this._ownerNode, (this._hoverValue - this.delay) / this.fadeInTime);
+            this._setAllVisibility(this._ownerNode, this._hovered ?

```

Comment:
Why not just using -delay here ?

Basically we could always use delay independently of fading in or out and you would not need to differentiate here, only during the setup ?

Reply:
Are you suggesting I create a local delay var and set it to `this._hovered ? this.delay : -this.fadeOutDelay`?  Or that I set `this.delay` to `-this.fadeOutDelay` in `fadeIn` if value is false.

Reply:
I was thinking a local delay = to either delay or -delay depending on hovered ?

Reply:
Okay.  I opted for the second option in my latest, but I can change it to a local if you prefer.  May be more readable since the decision would be closer to the code.